TripalCultivate/TripalCultivate-Phenotypes
# G bug.60 subsequent uploads

**Issue #60 When first file passes validation, subsequent uploads without page refresh do not trigger errors

## Motivation



After uploading a data file with passing validation result, attempt to re-upload a file (with error) does not trigger any validation error.

## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

This PR adds a check in formValidate() method to check for which stage the form is currently at and what triggering element caused to submit the form.

When the stage is greater than 1 (no longer in stage 1 where file upload is required) and the form triggering element is to Validate Data File, then the stage is reset to stage 1 and cache the value to update stage cacheing mechanism of the form.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Create test data in
- Genus (chado.organism)
- Project (chado.project)
- Stock (chado.stock)
2. Install and enable Tripal Cultivate Phenotypes module
3. Configure Genus Ontology Configuration for the genus created
4. Ensure allow new trait is set to true (checked), this will allow non-existent trait to be uploaded (to skip creating trait, method and unit).
5. Access mysite/admin/tripal/loaders/trpcultivate-phenotypes-share and use the Download a template file to create a tsv file.
6. Using test data, fill out the columns. Make sure that this file will pass validation by providing correct data values.
7. Proceed to upload stage and upload the file. With passing validation, the current stage is now either stage 2 or stage3.
8. Alter the data file so that it will fail validation (ie: empty column, invalid year or rep set to 0)
9. Enable or expand stage 1 and re-upload file.
This should trigger validation and stage is accounted for so that it shows the correct stage despite previous file upload moved the stages to either stage 2 or 3.
10. Repeat steps 6-9 while in stage 3.
